internal/mdns/mdns.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package mdns
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/whyrusleeping/mdns"
+)
+
+// Notifee is notified when a new peer is found.
+type Notifee interface {
+	HandlePeerFound(peer.AddrInfo)
+}
+
+// Service implements a mDNS service.
+type Service struct {
+	server     *mdns.Server
+	p2pHost    IDNetworker
+	serviceTag string
+	notifee    Notifee
+	logger     Logger
+	started    bool
+	stop       chan struct{}
+	done       chan struct{}
+}
+
+// NewService creates and returns a new mDNS service.
+func NewService(p2pHost IDNetworker, serviceTag string,
+	logger Logger, notifee Notifee) (service *Service) {
+	if serviceTag == "" {
+		serviceTag = "_ipfs-discovery._udp"
+	}
+
+	return &Service{
+		p2pHost:    p2pHost,
+		serviceTag: serviceTag,
+		notifee:    notifee,
+		logger:     logger,
+	}
+}
+
+// Start starts the mDNS service.
+func (s *Service) Start() (err error) {
+	ips, port := getMDNSIPsAndPort(s.p2pHost)
+
+	hostID := s.p2pHost.ID()
+
+	hostIDPretty := hostID.Pretty()
+	txt := []string{hostIDPretty}
+	mdnsService, err := mdns.NewMDNSService(hostIDPretty, s.serviceTag, "", "", int(port), ips, txt)
+	if err != nil {
+		return fmt.Errorf("creating mDNS service: %w", err)
+	}
+
+	server, err := mdns.NewServer(&mdns.Config{Zone: mdnsService})
+	if err != nil {
+		return fmt.Errorf("creating mDNS server: %w", err)
+	}
+	s.server = server
+
+	s.started = true
+	s.stop = make(chan struct{})
+	s.done = make(chan struct{})
```

Comment:
Should we put these lines in the constructor?

Reply:
These are here so the same service can be started -> stopped multiple times. I think we should design our services as such, what do you think?

Reply:
I wouldn't say this is a requirement in this case.  Would we ever start and stop this service?

Reply:
Just thinking in terms of service design in general, it should be restart-able. There is no reason to restart this service, but why make code worst if it can be better for the same amount of lines :wink: